nav2_costmap_2d/plugins/costmap_filters/speed_filter.cpp

```diff
@@ -0,0 +1,344 @@
+  speed_limit_(NO_SPEED_LIMIT), speed_limit_prev_(NO_SPEED_LIMIT)
+{
+}
+
+void SpeedFilter::initializeFilter(
+  const std::string & filter_info_topic)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  // Declare "speed_limit_topic" parameter specific to SpeedFilter only
+  declareParameter("speed_limit_topic", rclcpp::ParameterValue("speed_limit"));
+  node->get_parameter(name_ + "." + "speed_limit_topic", speed_limit_topic_);
+
+  filter_info_topic_ = filter_info_topic;
+  // Setting new costmap filter info subscriber
+  RCLCPP_INFO(
+    logger_,
+    "SpeedFilter: Subscribing to \"%s\" topic for filter info...",
+    filter_info_topic_.c_str());
+  filter_info_sub_ = node->create_subscription<nav2_msgs::msg::CostmapFilterInfo>(
+    filter_info_topic_, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&SpeedFilter::filterInfoCallback, this, std::placeholders::_1));
+
+  // Get global frame required for speed limit publisher
+  global_frame_ = layered_costmap_->getGlobalFrameID();
+
+  // Create new speed limit publisher
+  speed_limit_pub_ = node->create_publisher<nav2_msgs::msg::SpeedLimit>(
+    speed_limit_topic_, rclcpp::QoS(10));
+  speed_limit_pub_->on_activate();
+
+  // Reset speed conversion states
+  base_ = BASE_DEFAULT;
+  multiplier_ = MULTIPLIER_DEFAULT;
+  percentage_ = true;
+}
+
+void SpeedFilter::filterInfoCallback(
+  const nav2_msgs::msg::CostmapFilterInfo::SharedPtr msg)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!mask_sub_) {
+    RCLCPP_INFO(
+      logger_,
+      "SpeedFilter: Received filter info from %s topic.", filter_info_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      logger_,
+      "SpeedFilter: New costmap filter info arrived from %s topic. Updating old filter info.",
+      filter_info_topic_.c_str());
+    // Resetting previous subscriber each time when new costmap filter information arrives
+    mask_sub_.reset();
+  }
+
+  // Set base_/multiplier_ or use speed limit in % of maximum speed
+  base_ = msg->base;
+  multiplier_ = msg->multiplier;
+  if (msg->type == SPEED_FILTER_PERCENT) {
+    // Using speed limit in % of maximum speed
+    percentage_ = true;
+    RCLCPP_INFO(
+      logger_,
+      "SpeedFilter: Using expressed in a percent from maximum speed"
+      "speed_limit = %f + filter_mask_data * %f",
+      base_, multiplier_);
+  } else if (msg->type == SPEED_FILTER_ABSOLUTE) {
+    // Speed limit in absolute values is not implemented yet
+    RCLCPP_ERROR(logger_, "SpeedFilter: Speed limit in absolute values is not implemented yet");
+    throw std::runtime_error("Mode is not implemented yet");
```

Comment:
My bad, please `return` here an in L134 rather than `throw`. This is in a callback, I glanced over that, so there's nothing to catch this exception and it will just crash the costmap process. `return` to stop action and a nasty error log is enough for now.

Reply:
Fixed starting from https://github.com/ros-planning/navigation2/pull/2074/commits/34e694dbee7af9f805d2a7038b292466e44dc104#diff-d344e376ecbc8e5ecd5045731e13855c6d2a9a955cdaee9b31dddbbee7681466L132